src/data/polynomial/laurent.lean

```diff
@@ -0,0 +1,314 @@
+  (h_C_mul_T_Z : ∀ (n : ℕ) (a : R), M (C a * T (- n)) → M (C a * T (- n - 1))) :
+  M p :=
+begin
+  have A : ∀ {n : ℤ} {a : R}, M (C a * T n),
+  { assume n a,
+    apply n.induction_on,
+    { simpa only [T_zero, mul_one] using h_C a },
+    { exact λ m, h_C_mul_T m a },
+    { exact λ m, h_C_mul_T_Z m a } },
+  have B : ∀ (s : finset ℤ), M (s.sum (λ (n : ℤ), C (p.to_fun n) * T n)),
+  { apply finset.induction,
+    { convert h_C 0, simp only [finset.sum_empty, _root_.map_zero] },
+    { assume n s ns ih, rw finset.sum_insert ns,
+    { exact h_add A ih } } },
+  convert B p.support,
+  ext a,
+  simp_rw [← single_eq_C_mul_T, finset.sum_apply', single_apply, finset.sum_ite_eq'],
+  split_ifs with h h,
+  { refl },
+  { exact finsupp.not_mem_support_iff.mp h }
+end
+
+/--  To prove something about Laurent polynomials, it suffices to show that
+* the condition is closed under taking sums, and
+* it holds for monomials.
+-/
+@[elab_as_eliminator] protected lemma induction_on' {M : R[T;T⁻¹] → Prop} (p : R[T;T⁻¹])
+  (h_add     : ∀p q, M p → M q → M (p + q))
+  (h_C_mul_T : ∀(n : ℤ) (a : R), M (C a * T n)) :
+  M p :=
+begin
+  refine p.induction_on (λ a, _) h_add _ _;
+  try { exact λ n f Mf, h_C_mul_T _ f },
+  convert h_C_mul_T 0 a,
+  exact (mul_one _).symm,
+end
+
+lemma commute_T (n : ℤ) (f : R[T;T⁻¹]) : commute (T n) f :=
+f.induction_on' (λ p q Tp Tq, commute.add_right Tp Tq) $ λ m a,
+show T n * _ = _, by
+{ rw [T, T, ← single_eq_C, single_mul_single, single_mul_single, single_mul_single],
+  simp [add_comm] }
+
+@[simp]
+lemma T_mul (n : ℤ) (f : R[T;T⁻¹]) : T n * f = f * T n :=
+(commute_T n f).eq
+
+/--  `trunc : R[T;T⁻¹] →+ R[X]` maps a Laurent polynomial `f` to the polynomial whose terms of
+strictly positive degree coincide with the ones of `f`.  The coefficients of the terms of
+non-positive degree of `f` are summed up and used as constant term.
+
+`trunc` is a left-inverse to `polynomial.to_laurent`. -/
+def trunc : R[T;T⁻¹] →+ R[X] :=
+((to_finsupp_iso R).symm.to_add_monoid_hom).comp (map_domain.add_monoid_hom int.to_nat)
+
+@[simp]
+lemma trunc_C_mul_T (n : ℤ) (r : R) : trunc (C r * T n) = monomial n.to_nat r :=
+begin
+  apply (to_finsupp_iso R).injective,
+  -- the combination `rw ← single_eq_C_mul_T, simp [-single_eq_C_mul_T, trunc], refl`
+  -- closes the goal, but involves a non-terminal `simp`.
+  rw [← single_eq_C_mul_T, trunc, add_monoid_hom.coe_comp, function.comp_app,
+    map_domain.add_monoid_hom_apply, to_finsupp_iso_apply, map_domain_single, to_finsupp_iso_apply,
+    to_finsupp_monomial],
+  refl,
+end
+
+lemma _root_.polynomial.to_laurent_injective :
+  function.injective (polynomial.to_laurent : R[X] → R[T;T⁻¹]) :=
+begin
+  refine function.has_left_inverse.injective ⟨trunc, λ f, f.induction_on' _ _⟩,
+  { exact λ f g hf hg, by simp only [hf, hg, _root_.map_add] },
+  { exact λ n r, by simp only [monomial_eq_C_mul_T, trunc_C_mul_T, int.to_nat_coe_nat] }
+end
```

Comment:
You could extract a direct proof that `trunc` is the left-inverse to `polynomial.to_laurent` from this proof.

Reply:
Maybe, but when I was trying to prove injectivity, I felt that this was a better way of proceeding.

I feel that both results are useful, so I might as well prove this one first.

Would you prefer that I proved injectivity independently?

Reply:
Let me explain again; my point is that you can extract lines 240.75 - 242 (starting after the lambda) into their own lemma that states `trunc p.to_laurent = p`, which would make a good simp lemma